Bidi + Font Shaping Support

JimBobSquarePants opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/Fonts/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
<!-- A description of the changes proposed in the pull-request -->
Implementation of bidi layout support + glyph shaping. Includes a complete rewrite of the layout engine.

Information and reference code adapted from
https://docs.microsoft.com/en-us/typography/opentype/spec/
https://github.com/harfbuzz/harfbuzz
https://github.com/foliojs/fontkit
https://github.com/LayoutFarm/Typography

Fixes #150, Fixes #66

Also provides the basic for supporting #8, and #9

#### Implemented features.
- [x] Mixed Bidi layout support LTR, RTL, LTR + RTL
- [x] LayoutMode.Horizontal
- [x] LayoutMode.Vertical
- [x] TextAlign
- [x] CodePoint Script class detection
- [x] Glyph Shaping
- - [x] Default Shaper
- - [x] Arabic Shaper
- - [ ] Indic Shaper (_future_)
- - [ ] Hangul Shaper (_future_)
- - [ ] Hebrew Shaper (_future_)
- - [ ] Khmer Shaper (_future_)
- - [ ] Thai Shaper (_future_)
- - [ ] Myanmar Shaper (_future_)
- - [ ] Universal Shaper (_future_)
- - [x] Gsub Table
- - - [x] Gsub Format 1
- - - [x] Gsub Format 2
- - - [x] Gsub Format 3
- - - [x] Gsub Format 4
- - - [x] Gsub Format 5
- - - [x] Gsub Format 6
- - - [x] Gsub Format 7
- - - [x] Gsub Format 8
- - [x] GPos Table
- - - [x] GPos Format 1
- - - [x] GPos Format 2
- - - [x] GPos Format 3
- - - [x] GPos Format 4
- - - [x] GPos Format 5
- - - [x] GPos Format 6
- - - [x] GPos Format 7
- - - [x] GPos Format 8
- - - [x] GPos Format 9
<!-- Thanks for contributing to SixLabors.Fonts! -->

### Example Output

**Chinese Horizontal**

![horizontal-ltr](https://user-images.githubusercontent.com/385879/139051880-c5364094-538b-4c19-8510-00169b35907e.png)

**Chinese Vertical (_columns are read right to left; the text direction is top to bottom_)**

![vertical-rtl](https://user-images.githubusercontent.com/385879/139051884-2a0365a1-4a61-421c-b3c3-a5212d0be3c0.png)

**Combined emoji (_4 scalar values and 7 chars_)**

![👩🏽‍🚒a](https://user-images.githubusercontent.com/385879/139052673-e29aabf1-92da-4d52-bca7-ce4e313c902f.png)

**Arabic with complex ligatures**

![بِسْمِ ٱللَّهِ ٱلرَّحْمَٟنِ ٱلرَّحِيمِ](https://user-images.githubusercontent.com/385879/139052758-2f22964b-c620-404d-8673-bda6f7078162.png)

**Mixed English and Arabic.**

![English اَلْعَرَبِيَّةُ English](https://user-images.githubusercontent.com/385879/139052940-28ff138b-94e3-42b3-a24b-565fb75bcceb.png)

**Text alignment**

![😀A😀-left-bottom](https://user-images.githubusercontent.com/385879/139053224-0cd0442a-d86f-406a-8c18-5468acb30cc6.png)
![😀A😀-left-center](https://user-images.githubusercontent.com/385879/139053225-33557462-c8dc-453b-8073-4ad49b8ce738.png)
![😀A😀-left-top](https://user-images.githubusercontent.com/385879/139053230-047dcef1-58a3-4ce3-a61a-d88e290b1fd6.png)
![😀A😀-center-bottom](https://user-images.githubusercontent.com/385879/139053214-1902b12b-4db5-447d-8442-c687a701eab4.png)
![😀A😀-center-center](https://user-images.githubusercontent.com/385879/139053220-12f0a3b5-0c0f-4a22-b48b-62c6da9f528f.png)
![😀A😀-center-top](https://user-images.githubusercontent.com/385879/139053221-d09abe46-4dc0-4c9e-a5c8-de3be7f82670.png)
![😀A😀-right-bottom](https://user-images.githubusercontent.com/385879/139053232-8f837e25-5425-4894-be4a-57905ed79fca.png)
![😀A😀-right-center](https://user-images.githubusercontent.com/385879/139053234-e236556d-2424-43c8-b24d-c4391787ed8f.png)
![😀A😀-right-top](https://user-images.githubusercontent.com/385879/139053237-0dbb4d57-16c8-47bf-82d8-5fb144d8f84e.png)

**Word wrapping (_normal_)**

![THISIS~1](https://user-images.githubusercontent.com/385879/139053649-826b6586-3a6c-40c2-b673-249bdc96569d.PNG)

**Word wrapping (_break all_)**

![THISIS~1](https://user-images.githubusercontent.com/385879/139054136-85748ccb-6819-4c1b-9a7a-8f2f05be8636.PNG)

**Word wrapping (_keep all_)**

![THISIS~1](https://user-images.githubusercontent.com/385879/139054210-94cd5605-d838-4a70-ab7b-cdec436a5bd7.PNG)

JimBobSquarePants commented on src/SixLabors.Fonts/RendererOptions.cs:
```diff
@@ -173,7 +173,7 @@ internal AppliedFontStyle GetStyle(int index, int length)
             return new AppliedFontStyle
             {
-                Start = 0,
+                Start = index,
```
Found we were always using zero. Tests were incorrectly expecting that also.

JimBobSquarePants commented on src/SixLabors.Fonts/AppliedFontStyle.cs:
```diff
@@ -17,6 +17,8 @@ internal struct AppliedFontStyle
 
+        // TODO: This should accept all the glyphs in the range to
+        // allow complex substitutions.
         public GlyphMetrics[] GetGlyphLayers(CodePoint codePoint, ColorFontSupport colorFontOptions)
```
@tocsoft By my understanding this is our injection point for identifying the correct glyphs for each codepoint. 

It looks like we need to load the entire codepoint buffer though before laying things out in order to remap the glyph ids in context (_we could potentially remove codepoints_).  

I'm seeing though in the run that we load up a style for the entire buffer and never reach the end of that style even though we check per codepoint below. Am I reading that correctly?

JimBobSquarePants commented on tests/SixLabors.Fonts.Tests/GlyphTests.cs:
```diff
@@ -137,11 +137,14 @@ public void RenderWoff2Glyphs_EqualToTtfGlyphs()
             TextRenderer.RenderTextTo(rendererTtf, testStr, new RendererOptions(fontTtf)
             {
+                // TODO: Check what is different about the woff font.
+                ApplyKerning = false,
```
I'm not sure what is happening here. I did a comparison of the rendered output and the results are the same.

@brianpopow you any ideas?

![ttf](https://user-images.githubusercontent.com/385879/132083522-831cc511-590a-423c-80b5-46edba25a824.png)
![woff2](https://user-images.githubusercontent.com/385879/132083523-4415de84-f38a-4b18-9de7-e877cdf0dd2d.png)

brianpopow replied:
@JimBobSquarePants I have just tried this and the test passes. Do i understand this correct, that test fails for you?

JimBobSquarePants replied:
I’ve turned off kerning in the test. Turning it on leads to different vectors in the results for me.

brianpopow replied:
It seems the issue was introduced with d94e0b9e. 

I am not 100% sure yet, but i suspect the change to `AppliedFontStyle.cs`, could be the reason.

JimBobSquarePants replied:
Thanks yeah, massive rewrites to that section. I’ll be the ttf and woff2 fonts doing different things during glyph substitution but I would have thought the gsub table would be identical in both.

brianpopow replied:
yeah I think you are right, there is something different going on during the substitution. Not sure how to fix that yet, but this seems to be the reason for the difference.

JimBobSquarePants replied:
Woff isn't doing the substitution. The `I` character is being replaced in the TTF version.

brianpopow replied:
glad you found this

brianpopow replied:
@JimBobSquarePants I would like to change the test a bit and also test for the correct control points count. Is it ok if i commit this change to your branch?

JimBobSquarePants replied:
Yeah. Absolutely!

JimBobSquarePants commented:
Hey @tocsoft !! I got combined emojis working

  ![👩🏽‍🚒](https://user-images.githubusercontent.com/385879/132227813-25176e06-c8f8-4c61-8264-5821ac66df10.png)

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/Gsub/LookupType5SubTable.cs:
```diff
@@ -0,0 +1,359 @@
+            {
+                // TODO: Implement
+                // https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#example-7-contextual-substitution-format-1
+                return false;
```
@brianpopow I could do with your considerable spec parsing skills here if you have time. I can follow the spec to determine how to decode the tables however I can't make sense of how to actually implement the substitution. I've been trying t0 understand HarfBuzz but that's hieroglyphs to me.

brianpopow replied:
sure, cant promise I will understand it better, but I will take a look.

brianpopow replied:
there is alot to chew on here. It will take some time for me to understand that. I am currently trying to understand how harfbuzz does it, hope that will help here.

brianpopow commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/CoverageTable.cs:
```diff
@@ -0,0 +1,164 @@
+            for (int i = 0; i < this.records.Length; i++)
+            {
+                CoverageRangeRecord rec = this.records[i];
+                if (rec.StartGlyphId <= glyphId && glyphId <= rec.EndGlyphId)
```
Is this maybe a mistake here and the comparison should be flipped?

```suggestion
                if (rec.StartGlyphId >= glyphId && glyphId <= rec.EndGlyphId)
```

JimBobSquarePants replied:
Ah no, the glyph has to be within the start and end of the range (inclusive).

JimBobSquarePants commented:
There's gaps in some of the GSub table support which I could do with help with completing. Lookups 5,6, & 8 which I'm sure someone with an understanding  of C++ or Rust would be able to flesh out.

Complete implementations can be found here.
https://github.com/RazrFalcon/rustybuzz (Rust)
https://github.com/harfbuzz/harfbuzz (C++)

tocsoft commented on tests/SixLabors.Fonts.Tests/Tables/AdvancedTypographic/Gsub/GSubTableTests.cs:
```diff
@@ -0,0 +1,67 @@
+            string testStr = "ب";
+            var rendererTtf = new ColorGlyphRenderer();
+            bool applyKerning = true;
+            Vector2[] expectedControlPoints = new[]
```
for this test would it not make more sense to figure out what the expected glyph index should be and validate its the one tryign to be rendered rather than looking at control points? 

You can access the rendered glyph index using the `GlyphKeys` collection on  `ColorGlyphRenderer`

brianpopow replied:
yeah thats much better, glad you mentioned it before i did tests for all the other characters 😄

tocsoft commented on tests/SixLabors.Fonts.Tests/Tables/AdvancedTypographic/Gsub/GSubTableTests.cs:
```diff
@@ -60,8 +26,8 @@ public void RenderArabicCharacters_WithIsolatedForm_Works()
             // assert
-            Assert.Equal(expectedControlPoints.Length, rendererTtf.ControlPoints.Count);
-            Assert.True(rendererTtf.ControlPoints.SequenceEqual(expectedControlPoints));
+            Assert.Single(rendererTtf.GlyphKeys);
+            Assert.Equal(expectedGlyphIndex, rendererTtf.GlyphKeys[0].GlyphIndex);
```
```suggestion
            var glyphKey = Assert.Single(rendererTtf.GlyphKeys);
            Assert.Equal(expectedGlyphIndex, glyphKey.GlyphIndex);
```

makes it slightly easier to read in my opinion

brianpopow replied:
yes, I agree

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/FeatureTag.cs:
```diff
@@ -0,0 +1,820 @@
+        /// the user may need a glyph outside the context supported by the normal substitution, or the user may not know what feature produces the desired glyph.
+        /// Since many-to-one substitutions are not covered, ligatures would not appear in this table unless they were variant forms of another ligature.
+        /// </summary>
+        Aalt = 0x61616C74U,
```
We actually have a Tag struct that's designed to hold all features, langs, and all sorts.

`Tag.Parse("aalt")` would give you a tag with the `uint` Value.

brianpopow commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/GSubTable.cs:
```diff
@@ -0,0 +1,152 @@
+            collection.GetCodePointAndGlyphIds(index, out CodePoint codePoint, out int _, out IEnumerable<int> _);
+
+            ScriptListTable scriptListTable = this.ScriptList.Default();
+            Tag[] tags = UnicodeScriptTagMap.Instance[CodePoint.GetScript(codePoint)];
```
I noticed, that the `CodePoint.GetScript` method returns a wrong value for arabic characters. For example "ب", CodePoint 1576 (hex: 628) it returns `CaucasianAlbanian`. My understanding is, that this is because `UnicodeCategory.trie` was build with only latin characters in mind. Is that correct @JimBobSquarePants ?

JimBobSquarePants replied:
Hmmm no, must be a bug in my generation: There should be 1291 Arabic codepoints in there according to the input txt file it’s generated from. Will double check

JimBobSquarePants replied:
Fixed. Generator was fine I was just reading the wrong trie!! 🤣 I've added a few test ranges to confirm it's correct now for now but will flesh out in the future.

JimBobSquarePants commented:
@brianpopow @tocsoft 

@mhmd-azeez has kindly volunteered to help keep us right with the Arabic shaping output so if you have any questions just ping him.

brianpopow commented:
Here are some first rendering results for text written from right to left:

![farsi](https://user-images.githubusercontent.com/38701097/133286381-73788f53-4083-4483-a6a5-a9a8ecae0d89.png)

@mhmd-azeez does this look correct to you?

Gillibald commented:
> Ah…. In the GsubTable we were already able to determine the Script for each individual codepoint. I hoped we’d be able to work like that rather than a single script per text block. How would we mix scripts?

Text needs to be itemized so each script run can be shaped separately. That also comes into play when you enable certain font features for a portion of the text.

mhmd-azeez commented:
@brianpopow Yes. the correct forms are used and are properly connected. Can you please write it beside an English text to make sure the x-height of the glyphs are correct?

Also here is some Kurdish text to try:

نەما «خانی» هەتا چیڕۆکی عیشقی ئێمە داڕێژێ
دەنا هێشتا لە کوردستان دەژین «زین و مەم»ێکی تر

This also helps test multiline alignment of the text

brianpopow commented:
> @brianpopow Yes. the correct forms are used and are properly connected. Can you please write it beside an English text to make sure the x-height of the glyphs are correct?

@mhmd-azeez thanks a lot for your review and suggestions, I will give it a try tomorrow.

JimBobSquarePants commented:
@brianpopow I fixed up those GPos subtables and implemented the update methods. Those tests are still failing though. How did you determine what he correct output should be?

JimBobSquarePants commented:
>Text needs to be itemized so each script run can be shaped separately. That also comes into play when you enable certain font features for a portion of the text.

@Gillibald Yeah we do that now. Seems to be working well.

https://github.com/SixLabors/Fonts/pull/198/files#diff-db3514b13e65027012cf86809afb38a9b455bc2464a86a4c677dada95a3e5336R211-R241

JimBobSquarePants commented:
Ah actually @brianpopow , the failing test is related to kerning. We probably shouldn't be applying Kerning and GPOS. Will have to look into how the kerning table works.

JimBobSquarePants commented:
@mhmd-azeez Could you please post that sample without the line break? I'm having trouble getting it out of the browser in the correct format.  Maybe stick it in a code block?

mhmd-azeez commented:
@JimBobSquarePants I updated the sample and put it in a code block

JimBobSquarePants commented:
Whelp! Something's not right!

![نەما «خانی» هەتا چیڕۆکی عیشقی ئێمە داڕێژێ-دەنا هێشتا لە کوردستان دەژین «زین و مەم»ێکی تر](https://user-images.githubusercontent.com/385879/133404985-d29cdfe6-adab-45f7-bdf1-61b6722f1b1d.png)

Looks like we need to split the text into by lines first then reverse the glyphs within that line.

I don't know what is going on with the brackets though. That's an odd one. If I copy/paste one from the sample into a separate string the direction gets reversed to our output.

mhmd-azeez commented:
I will try to collect more test cases for you guys to try over the coming days

mhmd-azeez commented:
Numbers are also a pain because they are actually rendered from left to right!

``
لە ساڵی ١٢٣٤ ڕوویدا
``

![image](https://user-images.githubusercontent.com/16880059/133415180-0e99a2da-c840-461e-87f6-58172d7cba7b.png)

Fonts:
1. Times new roman
2. Calibri
3. Arabic typesetting
4. Segoe UI

JimBobSquarePants commented:
We’ll correctly detect the direction for numbers I think and render as expected. 

Brackets are an odd one though. They should ideally be substituted I think? I can’t seem to find relevant info though!

mhmd-azeez commented:
Yeah, brackets get reversed:

```
 [] {} ()
```
![image](https://user-images.githubusercontent.com/16880059/133416380-30befb7e-1c91-499d-9465-61d6dc1d6f32.png)

And here is some punctuation for good measure:
```
زۆرمان باس کرد؛ باسی ئاو، کەش، هەوای کوردستان.
```
![image](https://user-images.githubusercontent.com/16880059/133416569-e99930fa-be0b-431b-b8e4-3880aab914ca.png)

JimBobSquarePants commented:
Hmmm I wonder how that’s handled? We could maintain a list based on the Unicode spec but that seems super hacky.

mhmd-azeez commented:
This page has a little bit of info on these "mirrored" characters:
- https://www.w3.org/International/articles/inline-bidi-markup/uba-basics#mirroring
- https://www.unicode.org/Public/UCD/latest/ucd/BidiMirroring.txt
- https://unicode.org/reports/tr44/#Bidi_Mirroring_Glyph

brianpopow commented:
Mixing letters with numbers seems also not correct.

This is the string "لە ساڵی ١٢٣٤ ڕوویدا"

![mix_letters_and_characters](https://user-images.githubusercontent.com/38701097/133455620-978555ff-c265-456b-8ded-6bed93b90c64.png)

mhmd-azeez commented:
@brianpopow yup. `لە ساڵی` should be on the right and `ڕوویدا` Should be on the left.

JimBobSquarePants commented:
@brianpopow It's the bidi run order. My offset calc currently only works if the first run in a sequence is LTR. 

I need to have a proper think about this and figure out the correct order to do things. If anyone finds a simple workflow document as to the best approach please share.

JimBobSquarePants commented on src/SixLabors.Fonts/TextLayout2.cs:
```diff
@@ -0,0 +1,483 @@
+                text = text.TrimEnd(null);
+            }
+
+            IEnumerable<TextLine> textLines = ProcessText(text, options);
```
@tocsoft @brianpopow This is the beginning of a major rewrite of how we layout text separating shaping, breaking from the actual layout code.

The `IEnumerable<TextLine>` contains a collection `glyph-metrics|bidi-run|codepoint-offset` values which should give us enough information to layout bidirectional horizontal text (_vertical if we want also_).

I still have a few things to figure out but I plan to split layout explicitly for LTR and RTL lines choosing the initial direction based upon the directional value of the first bidi run. There are tricky things to consider though.

https://github.com/SixLabors/Fonts/pull/198#issuecomment-920087412 for example is a real head scratcher. Arabic numbers are actually LTR so there's 3 bidi runs in play.

JimBobSquarePants commented on src/SixLabors.Fonts/TextLayout.cs:
```diff
@@ -15,415 +18,902 @@ internal class TextLayout
+                // not actually present in the text, reverse any contiguous sequence of characters that are at that level or higher.
+                // https://unicode.org/reports/tr9/#L2
+
+                // TODO: Fix this. I'm reversing the wrong number of times.
```
@brianpopow @tocsoft if one of you could have a look at this for me please that'd be great. I've stuffed up and am reversing the glyphs to incorrect number of times for the mixed numbers/text.

I've been testing the first example using Segoe UI.
https://github.com/SixLabors/Fonts/pull/198#issuecomment-919887482

As you can see the glyph runs are in the correct sequence, it's just the actual glyph ordering.
![لە ساڵی ١٢٣٤ ڕوویدا](https://user-images.githubusercontent.com/385879/133933416-0b9c7280-c584-46c6-992f-816439346038.png)

brianpopow replied:
I think it looks ok now:

![bidi_test](https://user-images.githubusercontent.com/38701097/133941475-6323e9a3-54eb-4f34-a4d6-669d819b9a19.png)

brianpopow commented:
> Another test case to check: Mathematical equations
> 
> ```
> ١١س + ٨ج = ٨٥١
> ```
> 
> ![image](https://user-images.githubusercontent.com/16880059/134200925-4c516d33-b9c1-4051-804f-c738c4134f5e.png)
> 
> Font: Noto Naskh Arabic

@mhmd-azeez thanks for the suggestion. Here is how it looks:
![math_formula](https://user-images.githubusercontent.com/38701097/134226490-750aef59-b61b-4a0e-8127-2e860bf58910.png)

The orientation of the characters look correct, but the plus and equal signs are not drawn as a placeholder character. I need to investigate why, not sure yet.

tocsoft commented:
> The orientation of the characters look correct, but the plus and equal signs are not drawn as a placeholder character. I need to investigate why, not sure yet.

I would guess the glyphs are just not in the font, and the render above is using a fallback font for those 2 characters.

mhmd-azeez commented:
@tocsoft you're right! I looked at the font using https://fontdrop.info/ and it doesn't have `+` and `=` symbols. I was using LibreOfficer Writer and it seems like it just falls back to another font

brianpopow commented:
> That's an interesting point though, does the library support font fallbacks?
yes, we do.

I have tried a different font called nazli, which has + and = signs: 

![math_nazli](https://user-images.githubusercontent.com/38701097/134231269-e0f94ee1-1565-4e82-af9d-38c6def3e5a7.png)


Looks correct @mhmd-azeez , right?

mhmd-azeez commented:
@brianpopow yes, it is correct :+1:

JimBobSquarePants commented:
@brianpopow I've just pushed changes so that GPOS now filters for features. I _think_ it's correct but I was going off the FontKit implementation rather than HarfBuzz. I could do with a review of the changes just to be sure. 

I did find one oddity in the original implementation. hashsets were being assigned at offsets and queried at indexes which was working since we're not skipping codepoints but potentially they can be skipped which meant I had to move things around a little. This added some overhead to the lookups which I'd love to figure out a way to avoid since a lookup could now fail without them. I may have overengineered things though!

brianpopow commented:
> @brianpopow I've just pushed changes so that GPOS now filters for features. I _think_ it's correct but I was going off the FontKit implementation rather than HarfBuzz. I could do with a review of the changes just to be sure.
> 
> I did find one oddity in the original implementation. hashsets were being assigned at offsets and queried at indexes which was working since we're not skipping codepoints but potentially they can be skipped which meant I had to move things around a little. This added some overhead to the lookups which I'd love to figure out a way to avoid since a lookup could now fail without them. I may have overengineered things though!

I am currently working on tests for GSUB, there are still a few things which are not working as expected. See failing tests. I want to figure this out before i move on to GPOS. I will review the changes to GPOS when im done with this.

brianpopow commented:
> > @brianpopow I've just pushed changes so that GPOS now filters for features. I _think_ it's correct but I was going off the FontKit implementation rather than HarfBuzz. I could do with a review of the changes just to be sure.

@JimBobSquarePants commit 39ab42b looks good to me, cant spot anything obvious which should be changed.

I will continue with gathering test fonts for GPOS, so we can test if everything works out as intended. One thing i spotted so far:
While LookupType1SubTable, Format 1 does apply the offset, the rendered output looks cut off as if the bounding box is not moved:
![2120](https://user-images.githubusercontent.com/38701097/134499628-4c4c8a92-b5d5-41f8-ad11-4fa705475c21.png)

Cant spot the issue so far. The offset is applied to the bounding box.

Font is: `GposLookupType1Format1.ttf`
Teststring: `\u0015\u0014`

JimBobSquarePants commented:
Ok. I'll have a look. I need to gather some more info for the sub tables also that we don't have just now. Whether the Codepoint is a mark or base glyph. I should be able to add it using existing category data. We need it for GSub Format 4 also as we're substituting things we shouldn't.

JimBobSquarePants commented:
![--](https://user-images.githubusercontent.com/385879/134503525-de37f1db-dee1-431b-8244-ff8e1fe028e9.png)
@brianpopow is this what you expect?

brianpopow commented:
> ![--](https://user-images.githubusercontent.com/385879/134503525-de37f1db-dee1-431b-8244-ff8e1fe028e9.png)
> @brianpopow is this what you expect?

That looks much better! The 20 should be placed 200 units to the left. Its a bit hard to tell if that's the case. Maybe adding some more characters to the right could help visually.

JimBobSquarePants commented:
Yeah I'd say this is correct. 

Original
![--](https://user-images.githubusercontent.com/385879/134504715-88585b31-ab42-40f7-bb77-0565f97bf392.png)

Shifted
![-- - Copy](https://user-images.githubusercontent.com/385879/134504712-e74bd0f5-d144-4478-b619-aa0c9e69f770.png)

JimBobSquarePants commented:
@brianpopow Found an interesting issue with Gsub Format4 subtable - Ligature substitution. 

If I test the following string `first\n\n\n\nl` with Segoe UI of Open Type then the `i` following the `f` is removed.

I had a look at Harfbuzz to see if there was something simple I could implement but unfortunately I cannot make sense of it.
https://github.com/harfbuzz/harfbuzz/blob/9247d24d7f41f820d7ea79089b482410b927edc8/src/hb-ot-layout-gsubgpos.hh#L931

JimBobSquarePants commented:
Sorry @brianpopow my bad... It was Segoe UI with the OpenSans-Regular.ttf as a fallback. 

I've updated the code to prevent look for fallbacks among control characters which fixes that issue however.... If you try to write that text using OpenSans the substitution takes place and it actually crashes the drawing processor with an IndexOutOfRangeException!

JimBobSquarePants commented:
@brianpopow just did a big refactor to pass the current feature as the input sequence matcher should check against it. Unfortunately the Open Sans issue still exists but the API is much easier to work with now. Big speedup also. About 2X faster for a test run on the Ubuntu CI 6secs vs 12.

brianpopow commented:
> @brianpopow just did a big refactor to pass the current feature as the input sequence matcher should check against it. Unfortunately the Open Sans issue still exists but the API is much easier to work with now. Big speedup also. About 2X faster for a test run on the Ubuntu CI 6secs vs 12.

I will try to figure out the issue with Open Sans.

brianpopow commented:
> Unfortunately the Open Sans issue still exists 

@JimBobSquarePants can you explain how exactly you can trigger the issue with OpenSans? I tried this test string `first\n\n\n\nl`
with segeoui font and OpenSans-Regular as the fall back font. This does not trigger an exception.

JimBobSquarePants commented:
@brianpopow stick this code in the Program.cs file in the samples project.

```c#
RenderText(font2, "first\n\n\n\nl", pointSize: 20);
```

JimBobSquarePants commented:
I'm relegated currently to attaching a debugger to https://fontkit-demo.vercel.app/ and loading the font+text as I cannot get a local build of it to run. So far I can see a substitution is taking place with the same return. However I have no idea what happens next.

JimBobSquarePants commented:
@brianpopow I'm 100% confident now that the issue is with ImageSharp.Drawing.

FontKit returns exactly the same glyph which has Id 564. In the font this represents codepoint 0xFB01 which is a Latin small ligature `fi` 

![image](https://user-images.githubusercontent.com/385879/134691813-920fc564-c32e-44bd-a6bc-13d76195bb90.png)

If I directly reference my local copy of ImageSharp.Drawing from the sample program in the Fonts solution I can attach the debugger. The exception is thrown in `ActiveEdgeList.LeaveEdge(idx)`. If I disable the exception I get the following output.

![ﬁ](https://user-images.githubusercontent.com/385879/134691527-ade4c1dc-db1c-428f-a489-8f6fb486e3a4.png)

As you can see the bounding box is correct and the `f` portion is renderable before the issue occurs. I think we should move on and try to fix the issue in Drawing once this is merged. @antonfirsov I'll absolutely need your help then. The rasterizer is a black box for me.

JimBobSquarePants commented:
Bugger.... Wind that all back. I just created a test project using latest nightlies and I can render it!
![fi](https://user-images.githubusercontent.com/385879/134698427-b2577b14-b45d-4542-9890-36c5015377e8.png)

Now to compare the font metrics....

JimBobSquarePants commented:
I'm thoroughly confused now. I can get it to throw using the latest Fonts master but I simply cannot get it to throw using the latest from MyGet.

JimBobSquarePants commented:
It's the woff version that's the issue! Likely a bad conversion. If you load the woff version followed by the ttf version into the font collection it cannot tell the difference and ignores the ttf. Not sure what to do about that behaviour, better equality comparison I think.

![ﬁ](https://user-images.githubusercontent.com/385879/134712709-a7a51583-c6ae-4993-82f4-6118f9724344.png)

brianpopow commented:
> It's the woff version that's the issue! Likely a bad conversion. If you load the woff version followed by the ttf version into the font collection it cannot tell the difference and ignores the ttf. Not sure what to do about that behaviour, better equality comparison I think.
> 
> ![ﬁ](https://user-images.githubusercontent.com/385879/134712709-a7a51583-c6ae-4993-82f4-6118f9724344.png)

It has something to do with the substitution also. If you disable substitutions the error does not happen.

JimBobSquarePants commented:
>It has something to do with the substitution also. If you disable substitutions the error does not happen. 

Yep so the substitution replaces the individual `f` & `i` with the single `fi` glyph which in the woff font is buggy (_The substitution is correct though_). You can trigger the same exception with that font by simply trying to layout the single `fi` glyph without kerning enabled.

JimBobSquarePants commented:
@brianpopow I've gotten a little side tracked and am just doing a little bit of testing and fixing of the woff2 loader. 

I've substituted the Opens Sans font I have locally for the [v3 version](https://google-webfonts-helper.herokuapp.com/fonts/open-sans?subsets=cyrillic,cyrillic-ext,greek,greek-ext,hebrew,latin,latin-ext,vietnamese) so I could get the proper woff2 file rather than a generated one. I couldn't load the file at first as the header reader was telling me there were duplicates but after a little refactoring of the table loading code based on the [official source](https://github.com/google/woff2/blob/master/src/woff2_dec.cc) I was able to get it loading. FontCreator can load and render the glyph well enough but we only seem to be detecting half the control points (_49 vs 87 where all points are equal until the cutoff_) when compared to the ttf version. I suspect something is incomplete. I've encountered that a fair amount in the LayoutFarm source.

`\uFB01` TTF
![ﬁ-ttf](https://user-images.githubusercontent.com/385879/134773275-0a2434b8-d28c-47c1-aba2-645926b4ee01.png)

`\uFB01` WOFF2
![ﬁ-woff2](https://user-images.githubusercontent.com/385879/134773277-bf9fcbb1-6c58-46b2-b0d7-f06e4c6c1742.png)

Incidentally the woff font cannot be loaded by us at all. It complains about a backwards seek (_FontCreator opens it just fine_).

brianpopow commented:
> @brianpopow I've gotten a little side tracked and am just doing a little bit of testing and fixing of the woff2 loader.
> 
> I've substituted the Opens Sans font I have locally for the [v3 version](https://google-webfonts-helper.herokuapp.com/fonts/open-sans?subsets=cyrillic,cyrillic-ext,greek,greek-ext,hebrew,latin,latin-ext,vietnamese) so I could get the proper woff2 file rather than a generated one. I couldn't load the file at first as the header reader was telling me there were duplicates but after a little refactoring of the table loading code based on the [official source](https://github.com/google/woff2/blob/master/src/woff2_dec.cc) I was able to get it loading. 
> ...

@JimBobSquarePants I would suggest to create a separate branch to fix this woff loading issues. This PR is already quite large.
Maybe we can start of with a unittest, which reproduces the issue(s), something like `RenderWoff2Glyphs_EqualToTtfGlyphs`.

JimBobSquarePants commented:
>I would suggest to create a separate branch to fix this woff loading issues. 

@brianpopow Yep, completely agree.

brianpopow commented:
> I could get the proper woff2 file rather than a generated one. I couldn't load the file at first as the header reader was telling me there were duplicates but after a little refactoring of the table loading code based on the

@JimBobSquarePants how did you manage to get the woff2 file to load. When I skip the duplicate table entry, I get a brotli decompression error.

JimBobSquarePants commented:
@brianpopow [See here](https://github.com/SixLabors/Fonts/pull/209/files#diff-3ef827409aef86b4ef7d3dc976ca9d4cb912e412b2b69d9688adfe3b73a42a74R48) I just copied the way Google do it.

JimBobSquarePants commented:
@brianpopow Just updated GPOS LookupType4SubTable to what I think is the correct approach. I couldn't find a test though and I wasn't sure where you were getting the expected info from.

brianpopow commented:
> @brianpopow Just updated GPOS LookupType4SubTable to what I think is the correct approach. I couldn't find a test though and I wasn't sure where you were getting the expected info from.

@JimBobSquarePants there is no test yet. Usually i try to get it to look like harfbuzz and then write a test comparing the glyph rectangles to expected rectangles. 
There is a test font `GposLookupType4Format1.ttf` from harfbuzz.
Use the string `\u0012\u0013`

This should look like:
![1819_aftergpos](https://user-images.githubusercontent.com/38701097/135472027-b2564b9c-b722-4eee-bd87-6631fe8ee8f9.png)

Actually looks like:
![output](https://user-images.githubusercontent.com/38701097/135472053-0041865c-31f2-4dff-9676-8f306d16b059.png)

My understanding is, we need to move the glyph at `index` (the 19) to the baseGlyph position (the "not mark" glyph) and the apply the offsets. I could not figure out how to properly achieve that. Maybe we need a better example font. I think some arabic scripts use this feature to add symbols over character to describe how vowels are pronounced (they called **fatheh, kaßreh and sammeh**). Any chance @mhmd-azeez may know a font which uses the gpos [Mark-to-Base Attachment Positioning](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-4-mark-to-base-attachment-positioning-subtable) for that?

JimBobSquarePants commented:
Ah ok… So that’s the output after my changes?

brianpopow commented:
> Ah ok… So that’s the output after my changes?

yes it is

JimBobSquarePants commented:
Cool. If you don’t beat me to it I’ll have a look again tomorrow

JimBobSquarePants commented:
@brianpopow there you go. The original doesn't seem to account for vertical layout so I took that into consideration also.

![--](https://user-images.githubusercontent.com/385879/135554988-fd1de821-bf27-40f8-bb1c-4ec5872c4615.png)

Btw, dunno if you've seen this but it's damn useful
https://fontkit-demo.vercel.app/

brianpopow commented:
> @brianpopow there you go. The original doesn't seem to account for vertical layout so I took that into consideration also.
> 

Good job, that looks correct! I will add a test for that.

> Btw, dunno if you've seen this but it's damn useful https://fontkit-demo.vercel.app/
> 

I did not know that, looks very useful, thanks.

mhmd-azeez commented:
@brianpopow This is beyond my area of expertise so I am not sure. As you explained these diacritics are used to show how the word is supposed to be pronounced. Most modern Arabic texts don't have them though because they are too time consuming to add and native Arabic speakers can pronounce the text without needing the diacritics usually. Here is an example sentence in Arabic that uses [Fatheh](https://www.compart.com/en/unicode/U+064E), [Kasrah](https://www.compart.com/en/unicode/U+061A), [Sukun](https://www.compart.com/en/unicode/U+0652), [Kasratan](https://www.compart.com/en/unicode/U+064D), and [Shaddah + Damma](https://www.compart.com/en/unicode/U+FC61):

```
كَيْفَ تُمِدُّ جِسْمَكَ بِالْـماءِ بِطَريقَةٍ صَحِيحَةٍ؟
```

For comparison, here is the same sentence without the diacritics:

```
كيف تمد جسمك بالماء بطريقة صحيحة؟
```

Times New Roman:
![image](https://user-images.githubusercontent.com/16880059/135652493-b0109ae9-2814-448d-99fa-df77299b6486.png)

Calibri:
![image](https://user-images.githubusercontent.com/16880059/135652533-f19b2d57-af74-4732-8ed3-c4e72d6e47c9.png)

Arabic Typesetting:
![image](https://user-images.githubusercontent.com/16880059/135652624-565b30f9-8164-4f28-91b2-1de796347796.png)

[Othmani Font](http://legionfonts.com/fonts/othmani):
![image](https://user-images.githubusercontent.com/16880059/135652411-cc767a17-e319-4416-8d4c-98ff2c1a451e.png)

Note: There are other diacritics used in Arabic:
https://en.wikipedia.org/wiki/Arabic_diacritics

brianpopow commented:
@mhmd-azeez thanks for the examples, those are helpful to find a font with this feature!

brianpopow commented:
@JimBobSquarePants there is still a bug in loading the gpos tables, i guess its related to Lookup Type 5. Will look into it tomorrow.

brianpopow commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/LookupType6SubTable.cs:
```diff
@@ -0,0 +1,132 @@
+                GlyphShapingData prevData = collection.GetGlyphShapingData(prevIdx);
+                if (!CodePoint.IsMark(prevData.CodePoint))
+                {
+                    return false;
```
`\0012` should be a Mark in the font GposLookupType6Format1.ttf, but `CodePoint.IsMark` returns false here. Maybe checking if its a mark via CodePoint is not correct here.

JimBobSquarePants replied:
Cool. I’ll have a look tonight for you

JimBobSquarePants replied:
Looking now.

brianpopow replied:
if you load the font with Font Creator, you can see the `OT Class` of `\0012` to be Mark.

JimBobSquarePants replied:
Ok. So our `CodePoint.IsMark()` check is correct, however this is not what FontKit nor HarfBuzz are checking. They both maintain `LigatureId` and `LigatureComponent` properties for each glyph and check those. 

We need to add those properties to the `GlyphShapingData` type and also a `SetGlyphShapingData(index, data)` method to `GlyphPositioningCollection`. We'll have to ensure we look back through the FontKit source to see when those properties are ever set.

JimBobSquarePants replied:
We very likely need GDEF table support.
https://docs.microsoft.com/en-us/typography/opentype/spec/gdef

JimBobSquarePants replied:
Hold on..... Yep. It's a combining mark when you check its LineBreakClass..... I need to do some research to make the IsMark test more robust!

JimBobSquarePants replied:
Ok So I did a whole bunch of research and augmented the method with the following code to cover all available ranges however that then broke lookups 4 and 5.

https://en.wikipedia.org/wiki/Combining_character
https://unicodesymbols.fandom.com/wiki/Category:Combining_Mark_Set

```c#
public static bool IsMark(CodePoint codePoint)
{
    // https://en.wikipedia.org/wiki/Combining_character

    // Combining Diacritical Marks
    if (UnicodeUtility.IsInRangeInclusive(codePoint.value, 0x0300, 0x036F))
    {
        return true;
    }

    // Combining Diacritical Marks Extended
    if (UnicodeUtility.IsInRangeInclusive(codePoint.value, 0x1AB0, 0x1AFF))
    {
        return true;
    }

    // Combining Diacritical Marks Supplement
    if (UnicodeUtility.IsInRangeInclusive(codePoint.value, 0x1DC0, 0x1DFF))
    {
        return true;
    }

    // Combining Diacritical Marks for Symbols
    if (UnicodeUtility.IsInRangeInclusive(codePoint.value, 0x20D0, 0x20FF))
    {
        return true;
    }

    // Combining Half Marks
    if (UnicodeUtility.IsInRangeInclusive(codePoint.value, 0xFE20, 0xFE2F))
    {
        return true;
    }

    return IsCategoryMark(GetGeneralCategory(codePoint)) || GetLineBreakClass(codePoint) == LineBreakClass.CM;
}
```

I'm not sure what the next step is. Do we simply check the linebreak class in lookup 6?

brianpopow replied:
My understanding so far was, that we need to implement the GDEF table and get the info, if the glyph is a mark from there, right?

JimBobSquarePants replied:
Unfortunately so. Unicode support should be enough but it doesn’t seem to be.

JimBobSquarePants replied:
Fontkit cheats by storing multiple codepoints against a glyph following a removal and counting that as a mark. Theoretically we could do the same but it’s yet another array

brianpopow replied:
Ok, I will start then implementing the GDEF table

brianpopow replied:
so far it seems to work out as planed: I use now the GDEF table if present to get the glyph class otherwise CodePoint.IsMark in LookupType6SubTable

JimBobSquarePants replied:
Awesome! That was quick!

brianpopow commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/LookupType5SubTable.cs:
```diff
@@ -0,0 +1,151 @@
+                ushort markGlyphId = glyphId;
+                ushort ligGlyphId = baseGlyphId;
+
+                // TODO: figure out how to calculate the compIndex, see fontKit.
```
not sure howto fix this TODO here. Im not sure what the ligatureID is supposed to be. @JimBobSquarePants maybe you have an idea?

to quote from harfbuzz
```
We must now check whether the ligature ID of the current mark glyph
is identical to the ligature ID of the found ligature.  If yes, we
can directly use the component index.  If not, we attach the mark
glyph to the last component of the ligature.
```

Its `hb-ot-layout-gpos-table.hh` line 1202

JimBobSquarePants replied:
So, looking at FontKit the ligature Id would be set in a single places. GSUB  `LookupType4SubTable`

The default value (taken from the processor, the `GlyphSubstitutionCollection` in our instance) would be `1`.

![image](https://user-images.githubusercontent.com/385879/136019118-7a7640f6-6220-48d4-8450-fbe8d3e6cf29.png)

They seem to carry a lot of info around with the data right through the process.

![image](https://user-images.githubusercontent.com/385879/136019851-ea835ee7-2525-4089-9b1a-e8873b277083.png)

JimBobSquarePants commented:
@brianpopow I've just pushed the font from the issue mentioned in the ligature substitution table. https://bugzilla.gnome.org/show_bug.cgi?id=437633

I got it from here.
https://sourceforge.net/projects/al-anvar/

Compared to FontKit it looks like we've still got some work to do, Seems like a whole set of the glyphs are missing.
![image](https://user-images.githubusercontent.com/385879/136791208-1fcf9331-5b1f-4fb1-9c5f-d15c41f96010.png)

JimBobSquarePants commented:
I'll have a look tomorrow. Late here now. I've been doing some more testing of the ligature substitution table and there definitely seems to be nothing taking place there... (_real headscratcher that one_)

JimBobSquarePants commented:
@brianpopow I've written the basic functionality for a `SkippingGlyphIterator` type which is important for determining which glyphs to operate upon and match. Currently it's a bit awkward to implement as all our individual matching functions should be using it but they all operate independently. I suggest we implement a `Match` method in `AdvancedTypographicUtils` that accepts a `Func` that can be called from all our matching methods. FontKit and HarfBuzz to the same.

JimBobSquarePants commented:
@brianpopow I managed to get the GSUB feature application working with the iterator by refactoring to do the following. We still preserve the correct application order plus can now skip when required. However, this does not change the result.

```c#
foreach (Tag stageFeature in shaper.GetShapingStageFeatures())
{
    List<(Tag Feature, ushort Index, LookupTable LookupTable)> lookups = this.GetFeatureLookups(stageFeature, current);
    if (lookups.Count == 0)
    {
        continue;
    }

    int currentCount = collection.Count;
    SkippingGlyphIterator iterator = new(fontMetrics, collection, index, default);

    // Apply features in order.
    foreach ((Tag Feature, ushort Index, LookupTable LookupTable) featureLookup in lookups)
    {
        Tag feature = featureLookup.Feature;
        iterator.Reset(index, featureLookup.LookupTable.LookupFlags);

        while (iterator.Index < index + count)
        {
            List<TagEntry> glyphFeatures = collection.GetGlyphShapingData(iterator.Index).Features;

            // TODO: No LINQ.
            if (!glyphFeatures.Any(x => x.Tag == featureLookup.Feature && x.Enabled))
            {
                iterator.Next();
                continue;
            }

            featureLookup.LookupTable.TrySubstitution(fontMetrics, this, collection, featureLookup.Feature, iterator.Index, count - (iterator.Index - index));
            iterator.Next();

            // Account for substitutions changing the length of the collection.
            if (collection.Count != currentCount)
            {
                count = (ushort)(count - (currentCount - collection.Count));
                currentCount = collection.Count;
            }
        }
    }
}
```

Gillibald commented:
`Chinese Vertical (columns are read right to left; text direction is left to right)` -> `Chinese Vertical (columns are read right to left; text direction is top to bottom)`

JimBobSquarePants commented:
Thanks @Gillibald just a bad copy/paste will fix. We’re definitely doing the right thing

Gillibald commented:
You did a really great job here. So impressive. Hopefully, this will be adopted by many projects so this can improve even further. .Net is used everywhere so there should be first-class text processing without any extra native dependencies. Your work is the first starting point to make that possible.

antonfirsov reviewed (commented):
A bunch of random notes on the code without understanding what's happening.

JimBobSquarePants commented on src/SixLabors.Fonts/TextLayout.cs:
```diff
@@ -13,88 +16,529 @@ namespace SixLabors.Fonts
-        /// <param name="text">The text.</param>
-        /// <param name="options">The style.</param>
-        /// <returns>A collection of layout that describe all that's needed to measure or render a series of glyphs.</returns>
+        internal static TextLayout Default { get; set; } = new();
```
It's an odd one. `TextMeasurer` accepts the instance as a parameter and stores it as a private field. Not sure whether @tocsoft had plans to allow replacing parts.

JimBobSquarePants commented on src/SixLabors.Fonts/RendererOptions.cs:
```diff
@@ -82,34 +81,25 @@ public RendererOptions(Font font, float dpiX, float dpiY, Vector2 origin)
 
         /// <summary>
-        /// Gets or sets a value indicating whether [apply kerning].
+        /// Gets or sets a value indicating whether to apply kerning.
```
So I'm currently doing this wrong.

There's two types of kerning in an OpenType font. Historical True Type kerning (_kern_) tables and modern glyph positioning (_gpos_) tables.

We're actually turning off glyph positioning and substitution with this switch when we shouldn't be.  

We should probably be allowing GSUB and GPOS always and only toggling kerning in the kern table. The [spec](https://www.w3.org/TR/css-fonts-3/#font-kerning-prop)states the same.

There's a CSS option for turning off [common ligatures](https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant-ligatures) which we should probably adopt also.

antonfirsov replied:
Hm, I only understand half of this, but if we are not sure the property controls the right thing, maybe better to remove it from the public API for now, and open a tracking issue to figure it out later?

JimBobSquarePants replied:
Yep. I've done some research and know how this works now. I'll push a fix in a day.

JimBobSquarePants replied:
Pushed. We now match the CSS specification with a new `KerningMode` property on `RendererOptions`. Unfortunately this will cause runtime errors in ImageSharp.Drawing until we update the `TextOptions` to reflect this (due to copying from ROps to TOps to new ROps instance)

JimBobSquarePants commented on src/SixLabors.Fonts/TextLayout.cs:
```diff
@@ -105,325 +549,560 @@ public IReadOnlyList<GlyphLayout> GenerateLayout(ReadOnlySpan<char> text, Render
-                            // to it ??? TODO: Should I be doing this?
-                            finalWidth += tabStop;
-                        }
+        internal sealed class TextBox
```
It's a container for a collection of text lines, so just once per render.

JimBobSquarePants commented on src/SixLabors.Fonts/Unicode/BidiRun.cs:
```diff
@@ -1,30 +1,38 @@
 namespace SixLabors.Fonts.Unicode
 {
-    internal readonly struct BidiRun
+    internal readonly struct BidiRun : IEquatable<BidiRun>
```
Most of the heavy documentation was for my purpose. I've found few things as complicated as GSUB/GPOS.

JimBobSquarePants commented on src/SixLabors.Fonts/RendererOptions.cs:
```diff
@@ -134,58 +124,41 @@ public RendererOptions(Font font, float dpiX, float dpiY, Vector2 origin)
+        /// Gets or sets the layout mode for the text lines.
         /// </summary>
-        public Vector2 Origin { get; set; } = Vector2.Zero;
+        public LayoutMode LayoutMode { get; set; }
```
Good question....

You would never actually render an Arabic font vertically, the language is explicitly horizontal and right-to-left so you would always rotate the glyphs and render them sideways (_we don't do that automatically_).

The visual layout (order) of glyphs is determined by the bidirectional rules of each script detected from the codepoints contained within the input string. For vertical alignment there is some default substitution of known codepoints but then shaping (gsub/gpos) rules apply. Our implementation of these follows HarfBuzz which work well. Glyphs that should be positioned together are tightly bound and vertical layout would not break that.

`LayoutMode.VerticalLeftRight`, and `LayoutMode.VerticalRightLeft` are really an opt-in thing for users who want to render languages vertically that support them. 

`LayoutMode.HorizontalTopBottom` was just showing off 😉

JimBobSquarePants commented on src/SixLabors.Fonts/Buffer{T}.cs:
```diff
@@ -0,0 +1,61 @@
+    /// An disposable buffer that is backed by an array pool.
+    /// </summary>
+    /// <typeparam name="T">The type of buffer element.</typeparam>
+    internal struct Buffer<T> : IDisposable
```
Yep agreed. Need to remove the interface though and use compiler pattern matching for ref structs. Done

JimBobSquarePants commented on src/SixLabors.Fonts/TextLayout.cs:
```diff
@@ -105,325 +549,560 @@ public IReadOnlyList<GlyphLayout> GenerateLayout(ReadOnlySpan<char> text, Render
+            public TextDirection TextDirection() => this.TextLines[0][0].TextDirection;
+        }
+
+        internal sealed class TextLine
```
Depends on the number of lines rendered. Usually very small numbers.

JimBobSquarePants commented on samples/DrawWithImageSharp/Program.cs:
```diff
@@ -36,6 +36,26 @@ public static void Main(string[] args)
+
+            //RenderText(uiFont, "Soft\u00ADHyphen", pointSize: 72);
+            RenderText(bugzilla, "بِسْمِ ٱللَّهِ ٱلرَّحْمَٟنِ ٱلرَّحِيمِ", pointSize: 72);
+            return;
+            // TODO: Test and fix tomorrow. Gsub Lookup 4.
```
Yeah. I'll likely delete most of it during a cleanup though.

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/UnicodeScriptTagMap.cs:
```diff
@@ -0,0 +1,192 @@
+    /// Provides a map from Unicode <see cref="ScriptClass"/> to OTF <see cref="Tag"/>.
+    /// <see href="https://docs.microsoft.com/en-us/typography/opentype/spec/scripttags"/>
+    /// </summary>
+    internal sealed class UnicodeScriptTagMap : Dictionary<ScriptClass, Tag[]>
```
We use `TryGet` I believe so not currently.

antonfirsov replied:
It's easy to implement `TryGet`: We can store `null` in the array for values which are not in `UnicodeScriptTagMap.Instance` but for some reason defined in `ScriptClass`.

Also note that all current values are arrays of `Length == 1`, so maybe we can store `Tag?` instead of `Tag[]`?

antonfirsov replied:
Note that I don't expect all these nits to be resolved in the PR, but would be nice to add TODO notes or a tracking issue documenting the code quality / perf leftovers.

JimBobSquarePants replied:
I recall why this is an array now. Historically there's been changes to features over the years with some made obsolete that we still need to associate with a script tag. There's evidence of various substitutions scattered throughout the HarfBuzz code but I haven't gotten around to collating them.

JimBobSquarePants replied:
Turns out I already had a TODO for that in the code! 😄

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/TagEntry.cs:
```diff
@@ -0,0 +1,21 @@
+namespace SixLabors.Fonts.Tables.AdvancedTypographic
+{
+    [DebuggerDisplay("Tag: {Tag}, Enabled: {Enabled}")]
+    internal class TagEntry
```
I'll check.

JimBobSquarePants commented on src/SixLabors.Fonts/GlyphSubstitutionCollection.cs:
```diff
@@ -0,0 +1,180 @@
+        public int LigatureId { get; set; } = 1;
+
+        /// <inheritdoc />
+        public ReadOnlySpan<ushort> this[int index] => this.glyphs[this.offsets[index]].GlyphIds;
```
So....

A Glyph Id is an identifier for a glyph in a font. Previously Glyph Index and Glyph Id were used interchangeably but I believe that to be incorrect. Lookups for glyphs are performed against a given codepoint and there's no rule that states they should be stored sequentially (_Only that zero is always the fallback glyph_). 

There's more than one at a given index within the collection because the collection is actually more representative of a jagged array (_though we don't use that for API ease of use_). You can have more than one glyph representing a single codepoint following substitution (_e.g. emojis_)

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/AttachPoint.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    internal struct AttachPoint
+    {
+        public ushort[] PointIndices;
```
Oooh! I'll check this out!

JimBobSquarePants replied:
Ah yeah, this has something to do with font variations which we don't support yet.

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/LigatureAttachTable.cs:
```diff
@@ -0,0 +1,42 @@
+    /// The order of the records also corresponds to the writing direction — that is, the logical direction — of the text.
+    /// For text written left to right, the first component is on the left; for text written right to left, the first component is on the right.
+    /// </summary>
+    internal class LigatureAttachTable
```
Yeah, I'll have a look.

JimBobSquarePants replied:
Gonna leave for now. I'd like to get this in and focus on potential performance issues after.

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/AdvancedTypographic/CoverageTable.cs:
```diff
@@ -0,0 +1,148 @@
+    {
+        public abstract int CoverageIndexOf(ushort glyphId);
+
+        public abstract IEnumerator<ushort> GetEnumerator();
```
That's a hangover of reading the Typography source. It can be removed.

JimBobSquarePants replied:
Done

antonfirsov commented on src/SixLabors.Fonts/KerningMode.cs:
```diff
@@ -0,0 +1,27 @@
+    /// Kerning is the contextual adjustment of inter-glyph spacing.
+    /// This property controls metric kerning, kerning that utilizes adjustment data contained in the font.
+    /// </summary>
+    public enum KerningMode
```
❤️

JimBobSquarePants commented on src/SixLabors.Fonts/Tables/General/CMapTable.cs:
```diff
@@ -14,29 +16,63 @@ internal sealed class CMapTable : Table
 
         private static int GetPreferredPlatformOrder(PlatformIDs platform)
-            => PreferredPlatformOrder.TryGetValue(platform, out var order) ? order : int.MaxValue;
+            => PreferredPlatformOrder.TryGetValue(platform, out int order) ? order : int.MaxValue;
```
Done

brianpopow reviewed (approved):
I think its time to get this huge PR merged! 👍
